File: app/src/main/java/com/frama/miserend/hu/home/view/HomeScreenActivity.java
package com.frama.miserend.hu.home.view;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.view.MenuItem;
import android.view.View;

import com.frama.miserend.hu.R;
import com.frama.miserend.hu.base.FragmentHostActivity;
import com.frama.miserend.hu.database.dialog.DatabaseDialogCallback;
import com.frama.miserend.hu.database.dialog.DatabaseMissingDialogFragment;
import com.frama.miserend.hu.database.dialog.DatabaseUpdateAvailableDialogFragment;
import com.frama.miserend.hu.database.local.LocalDatabase;
import com.frama.miserend.hu.database.miserend.manager.DatabaseState;
import com.frama.miserend.hu.home.pages.churches.view.ChurchesFragment;
import com.frama.miserend.hu.home.pages.map.view.ChurchesMapFragment;
import com.frama.miserend.hu.home.pages.masses.view.MassesFragment;
import com.frama.miserend.hu.home.viewmodel.HomeViewModel;
import com.frama.miserend.hu.location.LocationManager;
import com.frama.miserend.hu.router.Router;
import com.frama.miserend.hu.search.SearchParams;
import com.frama.miserend.hu.search.searchbar.CustomSearchBar;
import com.frama.miserend.hu.search.suggestions.Suggestion;
import com.frama.miserend.hu.search.suggestions.advanced.AdvancedSearchSuggestion;
import com.frama.miserend.hu.search.suggestions.church.ChurchSuggestion;
import com.frama.miserend.hu.search.suggestions.city.CitySuggestion;
import com.frama.miserend.hu.search.suggestions.recent.RecentSearchSuggestion;
import com.frama.miserend.hu.search.suggestions.viewmodel.SuggestionViewModel;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Balazs on 2018. 02. 10..
 */

public class HomeScreenActivity extends FragmentHostActivity implements DatabaseDialogCallback {

    @Inject
    HomeViewModel viewModel;
    @Inject
    SuggestionViewModel suggestionViewModel;
    @Inject
    Router router;
    @Inject
    LocationManager locationManager;

    @BindView(R.id.search_bar)
    CustomSearchBar searchBar;
    @BindView(R.id.bottom_navigation)
    BottomNavigationView bottomNavigationView;
    @BindView(R.id.search_fader)
    View searchFader;

    ProgressDialog databaseDownloadingDialog;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);
        ButterKnife.bind(this);
        viewModel.getDatabaseState().observe(this, this::onDatabaseStateChanged);
        bottomNavigationView.setOnNavigationItemSelectedListener(this::onNavigationItemSelected);
        suggestionViewModel.getSuggestions().observe(this, this::onSuggestionsChanged);
        searchFader.setOnClickListener(view -> searchBar.close());
        searchBar.setSearchBarCallback(new CustomSearchBar.SearchBarCallback() {
            @Override
            public void onSearchTermChanged(String searchTerm) {
                suggestionViewModel.updateSuggestions(searchTerm);
            }

            @Override
            public void onSearchStateChanged(boolean enabled) {
                searchFader.setVisibility(enabled ? View.VISIBLE : View.GONE);
                if (enabled) {
                    suggestionViewModel.updateSuggestions("");
                }
            }

            @Override
            public void onSuggestionSelected(Suggestion suggestion) {
                if (suggestion instanceof ChurchSuggestion) {
                    router.showChurchDetails(((ChurchSuggestion) suggestion).getData());
                } else if (suggestion instanceof AdvancedSearchSuggestion) {
                    router.showAdvancedSearch();
                } else if (suggestion instanceof CitySuggestion) {
                    SearchParams searchParams = new SearchParams();
                    searchParams.setCity(((CitySuggestion) suggestion).getData());
                    router.showSearchResults(searchParams);
                } else if (suggestion instanceof RecentSearchSuggestion) {
                    SearchParams searchParams = new SearchParams(((RecentSearchSuggestion) suggestion).getData());
                    router.showSearchResults(searchParams);
                }
            }

            @Override
            public void onSearchConfirmed(String searchTerm) {
                suggestionViewModel.addRecentSearch(searchTerm);
                searchBar.close();
                router.showSearchResults(new SearchParams(searchTerm));
            }
        });
    }

    private void onSuggestionsChanged(List<Suggestion> suggestions) {
        searchBar.updateLastSuggestions(suggestions);
    }

    private void onDatabaseStateChanged(DatabaseState databaseState) {
        switch (databaseState) {
            case UP_TO_DATE:
                hideDownloadingDialog();
                showFragment(new ChurchesFragment());
                break;
            case UPDATE_AVAILABLE:
                hideDownloadingDialog();
                showDatabaseUpdateDialog();
                break;
            case NOT_FOUND:
                hideDownloadingDialog();
                showDatabaseMissingDialog();
                break;
            case DOWNLOADING:
                databaseDownloadingDialog = ProgressDialog.show(this, null, getString(R.string.dialog_db_downloading));
                databaseDownloadingDialog.setCancelable(false);
                break;
        }
    }

    private void hideDownloadingDialog() {
        if (databaseDownloadingDialog != null) {
            databaseDownloadingDialog.cancel();
        }
    }

    private void showDatabaseMissingDialog() {
        DialogFragment newFragment = DatabaseMissingDialogFragment.newInstance();
        newFragment.show(getSupportFragmentManager(), "dialog");
    }

    private void showDatabaseUpdateDialog() {
        DialogFragment newFragment = DatabaseUpdateAvailableDialogFragment.newInstance();
        newFragment.show(getSupportFragmentManager(), "dialog");
    }

    public void downloadDatabase() {
        viewModel.downloadDatabase();
    }

    private void showFragment(Fragment fragment) {
        getSupportFragmentManager().beginTransaction().replace(R.id.fragment_container, fragment).commit();
    }

    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_churches:
                showFragment(new ChurchesFragment());
                return true;
            case R.id.action_masses:
                showFragment(new MassesFragment());
                return true;
            case R.id.action_map:
                showFragment(new ChurchesMapFragment());
                return true;
            default:
                return false;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        locationManager.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (LocationManager.LOCATION_SETTINGS_REQUEST_CODE == requestCode) {
            locationManager.getLastKnownLocation();
        } else if (LocationManager.LOCATION_PERMISSION_REQUEST_CODE == requestCode) {
            locationManager.getLastKnownLocation();
        }
    }

    @Override
    public void onDownloadClicked() {
        downloadDatabase();
    }

    @Override
    public void onDontDownloadClicked(boolean dbMissing) {
        if (dbMissing) {
            finish();
        } else {
            showFragment(new ChurchesFragment());
        }
    }
}

File: app/src/main/java/com/frama/miserend/hu/home/view/HomeScreenActivity.kt
package com.frama.miserend.hu.home.view

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.view.MenuItem
import android.view.View

import com.frama.miserend.hu.R
import com.frama.miserend.hu.base.FragmentHostActivity
import com.frama.miserend.hu.database.dialog.DatabaseDialogCallback
import com.frama.miserend.hu.database.dialog.DatabaseMissingDialogFragment
import com.frama.miserend.hu.database.dialog.DatabaseUpdateAvailableDialogFragment
import com.frama.miserend.hu.database.miserend.manager.DatabaseState
import com.frama.miserend.hu.home.pages.churches.view.ChurchesFragment
import com.frama.miserend.hu.home.pages.map.view.ChurchesMapFragment
import com.frama.miserend.hu.home.pages.masses.view.MassesFragment
import com.frama.miserend.hu.home.viewmodel.HomeViewModel
import com.frama.miserend.hu.location.LocationPermissionHelper
import com.frama.miserend.hu.router.Router
import com.frama.miserend.hu.search.SearchParams
import com.frama.miserend.hu.search.searchbar.CustomSearchBar
import com.frama.miserend.hu.search.suggestions.Suggestion
import com.frama.miserend.hu.search.suggestions.advanced.AdvancedSearchSuggestion
import com.frama.miserend.hu.search.suggestions.church.ChurchSuggestion
import com.frama.miserend.hu.search.suggestions.city.CitySuggestion
import com.frama.miserend.hu.search.suggestions.recent.RecentSearchSuggestion
import com.frama.miserend.hu.search.suggestions.viewmodel.SuggestionViewModel

import javax.inject.Inject

import butterknife.BindView
import butterknife.ButterKnife

/**
 * Created by Balazs on 2018. 02. 10..
 */

class HomeScreenActivity : FragmentHostActivity(), DatabaseDialogCallback {

    @Inject
    internal var viewModel: HomeViewModel? = null
    @Inject
    internal var suggestionViewModel: SuggestionViewModel? = null
    @Inject
    internal var router: Router? = null
    @Inject
    internal var locationPermissionHelper: LocationPermissionHelper? = null

    @BindView(R.id.search_bar)
    internal var searchBar: CustomSearchBar? = null
    @BindView(R.id.bottom_navigation)
    internal var bottomNavigationView: BottomNavigationView? = null
    @BindView(R.id.search_fader)
    internal var searchFader: View? = null

    internal var databaseDownloadingDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        ButterKnife.bind(this)
        viewModel!!.databaseState.observe(this, Observer<DatabaseState> { this.onDatabaseStateChanged(it) })
        bottomNavigationView!!.setOnNavigationItemSelectedListener(OnNavigationItemSelectedListener { this.onNavigationItemSelected(it) })
        suggestionViewModel!!.suggestions.observe(this, Observer<List<Suggestion>> { this.onSuggestionsChanged(it) })
        searchFader!!.setOnClickListener { view -> searchBar!!.close() }
        searchBar!!.setSearchBarCallback(object : CustomSearchBar.SearchBarCallback {
            override fun onSearchTermChanged(searchTerm: String) {
                suggestionViewModel!!.updateSuggestions(searchTerm)
            }

            override fun onSearchStateChanged(enabled: Boolean) {
                searchFader!!.visibility = if (enabled) View.VISIBLE else View.GONE
                if (enabled) {
                    suggestionViewModel!!.updateSuggestions("")
                }
            }

            override fun onSuggestionSelected(suggestion: Suggestion<*>) {
                if (suggestion is ChurchSuggestion) {
                    router!!.showChurchDetails(suggestion.data)
                } else if (suggestion is AdvancedSearchSuggestion) {
                    router!!.showAdvancedSearch()
                } else if (suggestion is CitySuggestion) {
                    val searchParams = SearchParams()
                    searchParams.city = suggestion.data
                    router!!.showSearchResults(searchParams)
                } else if (suggestion is RecentSearchSuggestion) {
                    val searchParams = SearchParams(suggestion.data)
                    router!!.showSearchResults(searchParams)
                }
            }

            override fun onSearchConfirmed(searchTerm: String) {
                suggestionViewModel!!.addRecentSearch(searchTerm)
                searchBar!!.close()
                router!!.showSearchResults(SearchParams(searchTerm))
            }
        })
        if (!locationPermissionHelper!!.hasLocationPermission()) {
            locationPermissionHelper!!.showPermissionRequestPopup()
        }
    }

    private fun onSuggestionsChanged(suggestions: List<Suggestion<*>>) {
        searchBar!!.updateLastSuggestions(suggestions)
    }

    private fun onDatabaseStateChanged(databaseState: DatabaseState) {
        when (databaseState) {
            DatabaseState.UP_TO_DATE -> {
                hideDownloadingDialog()
                showFragment(ChurchesFragment())
            }
            DatabaseState.UPDATE_AVAILABLE -> {
                hideDownloadingDialog()
                showDatabaseUpdateDialog()
            }
            DatabaseState.DATABASE_CORRUPT -> {
                hideDownloadingDialog()
                showDatabaseCorruptDialog()
            }
            DatabaseState.NOT_FOUND -> {
                hideDownloadingDialog()
                showDatabaseMissingDialog()
            }
            DatabaseState.DOWNLOADING -> {
                databaseDownloadingDialog = ProgressDialog.show(this, null, getString(R.string.dialog_db_downloading))
                databaseDownloadingDialog!!.setCancelable(false)
            }
        }
    }

    private fun hideDownloadingDialog() {
        if (databaseDownloadingDialog != null) {
            databaseDownloadingDialog!!.cancel()
        }
    }

    private fun showDatabaseMissingDialog() {
        val newFragment = DatabaseMissingDialogFragment.newInstance(R.string.dialog_db_missing_title, R.string.dialog_db_missing_message)
        newFragment.show(supportFragmentManager, "dialog")
    }

    private fun showDatabaseCorruptDialog() {
        val newFragment = DatabaseMissingDialogFragment.newInstance(R.string.dialog_db_corrupt_title, R.string.dialog_db_corrupt_message)
        newFragment.show(supportFragmentManager, "dialog")
    }

    private fun showDatabaseUpdateDialog() {
        val newFragment = DatabaseUpdateAvailableDialogFragment.newInstance()
        newFragment.show(supportFragmentManager, "dialog")
    }

    fun downloadDatabase() {
        viewModel!!.downloadDatabase()
    }

    private fun showFragment(fragment: Fragment) {
        supportFragmentManager.beginTransaction().replace(R.id.fragment_container, fragment).commit()
    }

    fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_churches -> {
                showFragment(ChurchesFragment())
                return true
            }
            R.id.action_masses -> {
                showFragment(MassesFragment())
                return true
            }
            R.id.action_map -> {
                showFragment(ChurchesMapFragment())
                return true
            }
            else -> return false
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        viewModel!!.retryLocation()
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        if (LocationPermissionHelper.LOCATION_SETTINGS_REQUEST_CODE == requestCode) {
            viewModel!!.retryLocation()
        } else if (LocationPermissionHelper.LOCATION_PERMISSION_REQUEST_CODE == requestCode) {
            viewModel!!.retryLocation()
        }
    }

    override fun onDownloadClicked() {
        downloadDatabase()
    }

    override fun onDontDownloadClicked(dbMissing: Boolean) {
        if (dbMissing) {
            finish()
        } else {
            showFragment(ChurchesFragment())
        }
    }
}